File: UIKit and Animations/WWDC 2014/Building Adaptive Apps with UIKit/AdaptivePhotos An Adaptive Application/AdaptivePhotosAnAdaptiveApplication/AdaptivePhotos/AAPLProfileViewController.m
/*
 Copyright (C) 2014 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 
  A view controller that shows a user's profile.
  
 */

#import "AAPLProfileViewController.h"
#import "AAPLConversation.h"
#import "AAPLPhoto.h"
#import "AAPLUser.h"

@interface AAPLProfileViewController ()
@property (strong, nonatomic) UIImageView *imageView;
@property (strong, nonatomic) UILabel *nameLabel;
@property (strong, nonatomic) UILabel *conversationsLabel;
@property (strong, nonatomic) UILabel *photosLabel;

@property (copy, nonatomic) NSArray *constraints;
@end

@implementation AAPLProfileViewController

- (void)updateConstraintsForTraitCollection:(UITraitCollection *)collection
{
    NSDictionary *views = @{@"topLayoutGuide" : self.topLayoutGuide, @"imageView" : self.imageView, @"nameLabel" : self.nameLabel, @"conversationsLabel" : self.conversationsLabel, @"photosLabel" : self.photosLabel};
    NSMutableArray *newConstraints = [NSMutableArray array];
    if (collection.verticalSizeClass == UIUserInterfaceSizeClassCompact) {
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"|[imageView]-[nameLabel]-|" options:0 metrics:nil views:views]];
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"[imageView]-[conversationsLabel]-|" options:0 metrics:nil views:views]];
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"[imageView]-[photosLabel]-|" options:0 metrics:nil views:views]];
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[topLayoutGuide]-[nameLabel]-[conversationsLabel]-[photosLabel]" options:0 metrics:nil views:views]];
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[topLayoutGuide][imageView]|" options:0 metrics:nil views:views]];
        [newConstraints addObject:[NSLayoutConstraint constraintWithItem:self.imageView attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:self.view attribute:NSLayoutAttributeWidth multiplier:0.5 constant:0.0]];
    } else {
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"|[imageView]|" options:0 metrics:nil views:views]];
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"|-[nameLabel]-|" options:0 metrics:nil views:views]];
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"|-[conversationsLabel]-|" options:0 metrics:nil views:views]];
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"|-[photosLabel]-|" options:0 metrics:nil views:views]];
        [newConstraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"V:[topLayoutGuide]-[nameLabel]-[conversationsLabel]-[photosLabel]-20-[imageView]|" options:0 metrics:nil views:views]];
    }
	
    if (self.constraints) {
        [self.view removeConstraints:self.constraints];
    }
    self.constraints = newConstraints;
    [self.view addConstraints:self.constraints];
}

- (void)willTransitionToTraitCollection:(UITraitCollection *)newCollection withTransitionCoordinator:(id <UIViewControllerTransitionCoordinator>)coordinator
{
    [super willTransitionToTraitCollection:newCollection withTransitionCoordinator:coordinator];
    [coordinator animateAlongsideTransition:^(id <UIViewControllerTransitionCoordinatorContext> context) {
        [self updateConstraintsForTraitCollection:newCollection];
        [self.view setNeedsLayout];
    } completion:nil];
}



- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.title = NSLocalizedString(@"Profile", @"Profile");
    }
    return self;
}

- (void)loadView
{
    UIView *view = [[UIView alloc] init];
    view.backgroundColor = [UIColor whiteColor];
    
    self.imageView = [[UIImageView alloc] init];
    self.imageView.contentMode = UIViewContentModeScaleAspectFit;
    self.imageView.translatesAutoresizingMaskIntoConstraints = NO;
    [view addSubview:self.imageView];
    
    self.nameLabel = [[UILabel alloc] init];
    self.nameLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleHeadline];
    self.nameLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [view addSubview:self.nameLabel];
    
    self.conversationsLabel = [[UILabel alloc] init];
    self.conversationsLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
    self.conversationsLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [view addSubview:self.conversationsLabel];
    
    self.photosLabel = [[UILabel alloc] init];
    self.conversationsLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
    self.photosLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [view addSubview:self.photosLabel];
    
    self.view = view;
    [self updateUser];
    [self updateConstraintsForTraitCollection:self.traitCollection];
}

- (void)setUser:(AAPLUser *)user
{
    if (_user != user) {
        _user = user;
        if ([self isViewLoaded]) {
            [self updateUser];
        }
    }
}

- (NSString *)nameText
{
    return self.user.name;
}

- (NSString *)conversationsText
{
    return [NSString stringWithFormat:NSLocalizedString(@"%ld conversations", @"%ld conversations"), self.user.conversations.count];
}

- (NSString *)photosText
{
    NSUInteger photoCount = 0;
    for (AAPLConversation *conversation in self.user.conversations) {
        photoCount += conversation.photos.count;
    }
    return [NSString stringWithFormat:NSLocalizedString(@"%ld photos", @"%ld photos"), photoCount];
}

- (void)updateUser
{
    self.nameLabel.text = self.nameText;
    self.conversationsLabel.text = self.conversationsText;
    self.photosLabel.text = self.photosText;
    self.imageView.image = self.user.lastPhoto.image;
}

@end

File: UIKit and Animations/WWDC 2014/Building Adaptive Apps with UIKit/AdaptivePhotos An Adaptive Application/AdaptivePhotos_Swift/AdaptivePhotos_Swift/Controller/AAPLProfileViewController.swift
/*
Copyright (C) 2014 Apple Inc. All Rights Reserved.
See LICENSE.txt for this sample’s licensing information

Abstract:

A view controller that shows a user's profile.
*/

import UIKit

class AAPLProfileViewController: UIViewController {

	var user: AAPLUser! {
		didSet {
			if self.isViewLoaded {
				self.updateUser()
			}
		}
	}
	
	var nameText: String {
		return self.user.name
	}
	
	var conversationsText: String {
		return String(format: NSLocalizedString("%ld conversations", comment: "%ld conversations"), self.user.conversations.count)
	}
	
	var photosText: String {
		var photoCount = 0
		for conversation in self.user.conversations {
			photoCount += conversation.photos.count
		}
		return String(format: NSLocalizedString("%ld photos", comment: "%ld photos"), photoCount)

	}
	
	var imageView: UIImageView!
	var nameLabel: UILabel!
	var conversationsLabel: UILabel!
	var photosLabel: UILabel!
	
	var constraints: [NSLayoutConstraint]!
	
	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)
//	}
//	override func viewDidLoad() {
//		super.viewDidLoad()
//	}
//	override func loadView() {
//		super.loadView()
		
		let view = UIView()
		view.backgroundColor = .white
		
		self.imageView = UIImageView()
		self.imageView.contentMode = .scaleAspectFit
		self.imageView.translatesAutoresizingMaskIntoConstraints = false
		view.addSubview(self.imageView)
		
		self.nameLabel = UILabel()
		self.nameLabel.font = UIFont.preferredFont(forTextStyle: .headline)
		self.nameLabel.translatesAutoresizingMaskIntoConstraints = false
		view.addSubview(self.nameLabel)
		
		self.conversationsLabel = UILabel()
		self.conversationsLabel.font = UIFont.preferredFont(forTextStyle: .body)
		self.conversationsLabel.translatesAutoresizingMaskIntoConstraints = false
		view.addSubview(self.conversationsLabel)
		
		self.photosLabel = UILabel()
		self.conversationsLabel.font = UIFont.preferredFont(forTextStyle: .body)
		self.photosLabel.translatesAutoresizingMaskIntoConstraints = false
		view.addSubview(self.photosLabel)
		
		self.view = view
		self.updateUser()
		self.updateConstraint(for: self.traitCollection)
	}

	func updateConstraint(for collection: UITraitCollection) {
		
		let views = ["topLayoutGuide" : self.topLayoutGuide, "imageView" : self.imageView, "nameLabel" : self.nameLabel, "conversationsLabel" : self.conversationsLabel, "photosLabel" : self.photosLabel] as [String : Any]
		
		var newConstraints: [NSLayoutConstraint] = []
		
		if collection.verticalSizeClass == .compact {
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "|[imageView]-[nameLabel]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "[imageView]-[conversationsLabel]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "[imageView]-[photosLabel]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "V:|[topLayoutGuide]-[nameLabel]-[conversationsLabel]-[photosLabel]", options: NSLayoutFormatOptions(), metrics: nil, views: views))
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "V:|[topLayoutGuide][imageView]|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
			newConstraints.append(NSLayoutConstraint(item: self.imageView, attribute: .width, relatedBy: .equal, toItem: self.view, attribute: .width, multiplier: 0.5, constant: 0.0))
		}
		else {
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "|[imageView]|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "|-[nameLabel]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "|-[conversationsLabel]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "|-[photosLabel]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
			newConstraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "V:[topLayoutGuide]-[nameLabel]-[conversationsLabel]-[photosLabel]-20-[imageView]|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		}
		
		if self.constraints != nil {
			self.view.removeConstraints(self.constraints)
		}
		self.constraints = newConstraints
		self.view.addConstraints(self.constraints!)
		
	}
	
	override func willTransition(to newCollection: UITraitCollection, with coordinator: UIViewControllerTransitionCoordinator) {
		super.willTransition(to: newCollection, with: coordinator)
		
		coordinator.animate(alongsideTransition: { (context) in
			self.updateConstraint(for: newCollection)
			self.view.setNeedsLayout()

		}, completion: nil)
	}
	
	override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
		super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
		
		self.title = NSLocalizedString("Profile", comment: "Profile")
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	func updateUser() {
		self.nameLabel.text = self.nameText
		self.conversationsLabel.text = self.conversationsText
		self.photosLabel.text = self.photosText
		
		self.imageView.image = self.user.lastPhoto.image
	}
	
}
